talpid-core/src/dns/windows.rs

```diff
@@ -0,0 +1,211 @@
+#[derive(err_derive::Error, Debug)]
+#[error(no_from)]
+pub enum Error {
+    /// Failure to obtain an interface LUID given an alias.
+    #[error(display = "Failed to obtain LUID for the interface alias")]
+    InterfaceLuidError(#[error(source)] io::Error),
+
+    /// Failure to obtain an interface index.
+    #[error(display = "Failed to obtain index of the interface")]
+    InterfaceIndexError(#[error(source)] io::Error),
+
+    /// Failure to spawn netsh subprocess.
+    #[error(display = "Failed to spawn 'netsh'")]
+    SpawnNetsh(#[error(source)] io::Error),
+
+    /// Failure to spawn netsh subprocess.
+    #[error(display = "Failed to obtain system directory")]
+    GetSystemDir(#[error(source)] io::Error),
+
+    /// Failure to write to stdin.
+    #[error(display = "Failed to write to stdin for 'netsh'")]
+    NetshInput(#[error(source)] io::Error),
+
+    /// Failure to wait for netsh result.
+    #[error(display = "Failed to wait for 'netsh'")]
+    WaitNetsh(#[error(source)] io::Error),
+
+    /// netsh returned a non-zero status.
+    #[error(display = "'netsh' returned an error: {:?}", _0)]
+    NetshError(Option<i32>),
+
+    /// netsh did not return in a timely manner.
+    #[error(display = "'netsh' took too long to complete")]
+    NetshTimeout,
+}
+
+pub struct DnsMonitor {
+    current_index: Option<u32>,
+}
+
+impl super::DnsMonitorT for DnsMonitor {
+    type Error = Error;
+
+    fn new() -> Result<Self, Error> {
+        Ok(DnsMonitor {
+            current_index: None,
+        })
+    }
+
+    fn set(&mut self, interface: &str, servers: &[IpAddr]) -> Result<(), Error> {
+        let interface_luid = luid_from_alias(interface).map_err(Error::InterfaceLuidError)?;
+        let interface_index =
+            index_from_luid(&interface_luid).map_err(Error::InterfaceIndexError)?;
+
+        self.current_index = Some(interface_index);
+
+        let mut ipv4_index = 0;
+        let mut ipv6_index = 0;
+
+        let mut netsh_input = String::new();
+
+        for server in servers {
+            let index;
+
+            if server.is_ipv4() {
+                index = ipv4_index;
+                ipv4_index += 1;
+            } else {
+                index = ipv6_index;
+                ipv6_index += 1;
+            };
+
+            if index == 0 {
+                netsh_input.push_str(&create_netsh_set_command(interface_index, server));
+            } else {
+                netsh_input.push_str(&create_netsh_add_command(interface_index, server));
+            }
+        }
+
+        if ipv4_index == 0 {
+            netsh_input.push_str(&create_netsh_flush_command(interface_index, IpVersion::V4));
+        }
+        if ipv6_index == 0 {
+            netsh_input.push_str(&create_netsh_flush_command(interface_index, IpVersion::V6));
+        }
+
+        run_netsh_with_timeout(netsh_input, NETSH_TIMEOUT)?;
+
+        Ok(())
+    }
+
+    fn reset(&mut self) -> Result<(), Error> {
```

Comment:
Calling this is actively bad when we're just doing it prior to destroying the tunnel interface. Should we add another method that defaults to calling `reset` but that is no-op for most DNS monitor implementations? I'd want to use that whenever destroying the tunnel interface is sufficient to achieve the same result.